toolchain/check/cpp_thunk.cpp

```diff
@@ -0,0 +1,369 @@
+                  llvm::SmallVector<bool> > {
+  unsigned num_params = callee_function_decl.getNumParams();
+  llvm::SmallVector<clang::QualType> thunk_param_types;
+  thunk_param_types.reserve(num_params);
+  llvm::SmallVector<bool> param_type_changed;
+  param_type_changed.reserve(num_params);
+
+  for (const clang::ParmVarDecl* callee_param :
+       callee_function_decl.parameters()) {
+    clang::QualType param_type = callee_param->getType();
+    bool is_scalar_type = IsScalarType(ast_context, param_type);
+    if (!is_scalar_type) {
+      clang::QualType pointer_type = ast_context.getPointerType(param_type);
+      param_type = ast_context.getAttributedType(
+          clang::NullabilityKind::NonNull, pointer_type, pointer_type);
+    }
+    param_type_changed.push_back(!is_scalar_type);
+    thunk_param_types.push_back(param_type);
+  }
+
+  return {thunk_param_types, param_type_changed};
+}
+
+// Returns the thunk parameters using the callee function parameter identifiers.
+static auto BuildThunkParameters(
+    clang::ASTContext& ast_context,
+    const clang::FunctionDecl& callee_function_decl,
+    clang::FunctionDecl* thunk_function_decl)
+    -> llvm::SmallVector<clang::ParmVarDecl*> {
+  clang::SourceLocation clang_loc = callee_function_decl.getLocation();
+
+  unsigned num_params = thunk_function_decl->getNumParams();
+  CARBON_CHECK(callee_function_decl.getNumParams() == num_params);
+
+  const auto* thunk_function_proto_type =
+      thunk_function_decl->getFunctionType()->getAs<clang::FunctionProtoType>();
+
+  llvm::SmallVector<clang::ParmVarDecl*> thunk_params;
+  thunk_params.reserve(num_params);
+  for (unsigned i = 0; i < num_params; ++i) {
+    clang::ParmVarDecl* thunk_param = clang::ParmVarDecl::Create(
+        ast_context, thunk_function_decl, clang_loc, clang_loc,
+        callee_function_decl.getParamDecl(i)->getIdentifier(),
+        thunk_function_proto_type->getParamType(i), nullptr, clang::SC_None,
+        nullptr);
+    thunk_params.push_back(thunk_param);
+  }
+  return thunk_params;
+}
+
+// Returns the thunk function declaration given the callee function and the
+// thunk parameter types.
+static auto CreateThunkFunctionDecl(
+    clang::ASTContext& ast_context,
+    const clang::FunctionDecl& callee_function_decl,
+    llvm::ArrayRef<clang::QualType> thunk_param_types) -> clang::FunctionDecl* {
+  clang::SourceLocation clang_loc = callee_function_decl.getLocation();
+
+  clang::IdentifierInfo& identifier_info = ast_context.Idents.get(
+      callee_function_decl.getNameAsString() + "__carbon_thunk");
+
+  const clang::FunctionType* callee_function_type =
+      callee_function_decl.getFunctionType();
+  CARBON_CHECK(callee_function_type->isFunctionProtoType());
+
+  clang::QualType thunk_function_type = ast_context.getFunctionType(
+      callee_function_decl.getReturnType(), thunk_param_types,
+      callee_function_type->getAs<clang::FunctionProtoType>()
+          ->getExtProtoInfo());
```

Comment:
```suggestion
  const auto* callee_function_type =
      callee_function_decl.getFunctionType()->castAs<clang::FunctionProtoType>();

  clang::QualType thunk_function_type = ast_context.getFunctionType(
      callee_function_decl.getReturnType(), thunk_param_types,
      callee_function_type->getExtProtoInfo());
```
We can avoid doing the `FunctionType` -> `FunctionProtoType` conversion twice here. (`castAs` includes the assert.)

Reply:
Done.